cards/git-realm/writer.ts

```diff
@@ -0,0 +1,401 @@
+import { Repository, RemoteConfig, GitConflict, UnknownObjectId } from './lib/git';
+import { FileNotFound, OverwriteRejected } from './lib/tree';
+
+import { todo } from './lib/todo-any';
+
+import crypto from 'crypto';
+import Change from './lib/change';
+import os from 'os';
+import process from 'process';
+import CardstackError from '@cardstack/core/error';
+import { dir as mkTmpDir } from 'tmp-promise';
+import { MetaObject, SingleResourceDoc } from 'jsonapi-typescript';
+import { extractSettings } from './lib/git-settings';
+// import { PrimaryData } from 'jsonapi-typescript';
+
+import { Writer } from '@cardstack/core/writer';
+import { Session } from '@cardstack/core/session';
+import { UpstreamDocument, UpstreamIdentity, upstreamIdToCardId, upstreamIdToString } from '@cardstack/core/document';
+import { inject } from '@cardstack/hub/dependency-injection';
+import { AddressableCard } from '@cardstack/core/card';
+
+// import { Session } from '@cardstack/core/session';
+// import { UpstreamDocument, UpstreamIdentity } from '@cardstack/core/document';
+// import { inject } from '@cardstack/hub/dependency-injection';
+// import { AddressableCard } from '@cardstack/core/card';
+// import CardstackError from '@cardstack/core/error';
+
+// let counter = 0;
+
+// export default class EphemeralWriter implements Writer {
+//   ephemeralStorage = inject('ephemeralStorage');
+
+//   constructor(private realmCard: AddressableCard) {}
+
+//   async create(_session: Session, doc: UpstreamDocument, upstreamId: UpstreamIdentity | null) {
+//     let id = upstreamId ?? String(counter++);
+//     let saved = this.ephemeralStorage.store(doc, id, this.realmCard.csId);
+//     return { saved: saved!, id };
+//   }
+
+//   async update(_session: Session, id: UpstreamIdentity, doc: UpstreamDocument) {
+//     let version = doc.jsonapi.data.meta?.version;
+//     if (version == null) {
+//       throw new CardstackError('missing required field "meta.version"', {
+//         status: 400,
+//         source: { pointer: '/data/meta/version' },
+//       });
+//     }
+
+//     return this.ephemeralStorage.store(doc, id, this.realmCard.csId, String(version))!;
+//   }
+
+//   async delete(_session: Session, id: UpstreamIdentity, version: string | number) {
+//     this.ephemeralStorage.store(null, id, this.realmCard.csId, version);
+//   }
+// }
+
+import stringify from 'fast-json-stable-stringify';
+
+const defaultBranch = 'master';
+
+function getType(document: UpstreamDocument) {
+  return document.jsonapi.data.type;
+}
+
+function getId(document: UpstreamDocument) {
+  return document.jsonapi.data.id;
```

Comment:
let's use document.canonicalURL()--it's simpler logic wise and more straight forward/ consistent with what we are doing else where.
```suggestion
  return document.canonicalURL;
```

Reply:
Doh, I think I can you a bad suggestion here: I was mistakingly thinking `document` was an instance of `Card`. I see it's not. what you had before is ok.